How to fix syb for GHC 8.2

While testing some packages with GHC 8.2 recently, I noticed that `syb` was broken:

```
src/Data/Generics/Instances.hs:74:10: error:
    • Illegal instance declaration for ‘Data TypeRep’
        (All instance types must be of the form (T t1 ... tn)
         where T is not a synonym.
         Use TypeSynonymInstances if you want to disable this.)
    • In the instance declaration for ‘Data TypeRep’
   |
74 | instance Data TypeRep where
   |          ^^^^^^^^^^^^
```

"Fixing" this issue is a simple matter of applying this patch...

```diff
diff --git a/src/Data/Generics/Instances.hs b/src/Data/Generics/Instances.hs
index 250f049..562ddb1 100644
--- a/src/Data/Generics/Instances.hs
+++ b/src/Data/Generics/Instances.hs
@@ -70,10 +70,12 @@ myMkNoRepType = mkNorepType
 -- Instances of abstract datatypes (6)
 ------------------------------------------------------------------------------
 
+#if __GLASGOW_HASKELL__ < 801
 instance Data TypeRep where
   toConstr _   = error "toConstr"
   gunfold _ _  = error "gunfold"
   dataTypeOf _ = myMkNoRepType "Data.Typeable.TypeRep"
+#endif
 
 
 ------------------------------------------------------------------------------
```

...but I'm not sure if this is how you'd intend to fix it. The problem is that as of GHC 8.2, `TypeRep` doesn't really exist as it used to anymore. The `Type.Reflection` module provides these two datatypes:

```haskell
data TypeRep (a :: k) where ...
data SomeTypeRep where
     SomeTypeRep :: forall k (a :: k). !(TypeRep a) -> SomeTypeRep
```

And in `Data.Typeable`, it provides the following (somewhat backwards-compatible) type synonym:

```haskell
type TypeRep = SomeTypeRep
```

So one option would be replacing the old `Data TypeRep` instance with two new instances on GHC 8.2+ only:

```haskell
instance (Typeable k, Typeable a) => Data (TypeRep (a :: k)) where ...
instance Data SomeTypeRep where ...
```

But to be honest, I really don't know what `syb` is even using that orphan `Data TypeRep` instance for (after all, you can comment it out and `syb` still compiles!), so I don't know if you'd even find the additions worthwhile.

What do you think?

Eh... I don't know. This instance probably made sense when Data and
Typeable were very tied together; not sure we need it now. I'd be ok with
removing it after 8.0.1.

On 28 February 2017 at 14:19, Ryan Scott <notifications@github.com> wrote:

> While testing some packages with GHC 8.2 recently, I noticed that syb was
> broken:
>
> src/Data/Generics/Instances.hs:74:10: error:
>     • Illegal instance declaration for ‘Data TypeRep’
>         (All instance types must be of the form (T t1 ... tn)
>          where T is not a synonym.
>          Use TypeSynonymInstances if you want to disable this.)
>     • In the instance declaration for ‘Data TypeRep’
>    |
> 74 | instance Data TypeRep where
>    |          ^^^^^^^^^^^^
>
> "Fixing" this issue is a simple matter of applying this patch...
>
> diff --git a/src/Data/Generics/Instances.hs b/src/Data/Generics/Instances.hs
> index 250f049..562ddb1 100644--- a/src/Data/Generics/Instances.hs+++ b/src/Data/Generics/Instances.hs@@ -70,10 +70,12 @@ myMkNoRepType = mkNorepType
>  -- Instances of abstract datatypes (6)
>  ------------------------------------------------------------------------------
>  +#if __GLASGOW_HASKELL__ < 801
>  instance Data TypeRep where
>    toConstr _   = error "toConstr"
>    gunfold _ _  = error "gunfold"
>    dataTypeOf _ = myMkNoRepType "Data.Typeable.TypeRep"+#endif
>
>
>  ------------------------------------------------------------------------------
>
> ...but I'm not sure if this is how you'd intend to fix it. The problem is
> that as of GHC 8.2, TypeRep doesn't really exist as it used to anymore.
> The Type.Reflection module provides these two datatypes:
>
> data TypeRep (a :: k) where ...data SomeTypeRep where
>      SomeTypeRep :: forall k (a :: k). !(TypeRep a) -> SomeTypeRep
>
> And in Data.Typeable, it provides the following (somewhat
> backwards-compatible) type synonym:
>
> type TypeRep = SomeTypeRep
>
> So one option would be replacing the old Data TypeRep instance with two
> new instances on GHC 8.2+ only:
>
> instance (Typeable k, Typeable a) => Data (TypeRep (a :: k)) where ...instance Data SomeTypeRep where ...
>
> But to be honest, I really don't know what syb is even using that orphan Data
> TypeRep instance for (after all, you can comment it out and syb still
> compiles!), so I don't know if you'd even find the additions worthwhile.
>
> What do you think?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dreixel/syb/issues/14>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZABPR8ckLtPOqxQXh6BUd-muY1RWzzks5rhCzkgaJpZM4MOczx>
> .
>

--- a/src/Data/Generics/Instances.hs
+++ b/src/Data/Generics/Instances.hs
@@ -70,10 +70,12 @@ myMkNoRepType = mkNorepType
 -- Instances of abstract datatypes (6)
 ------------------------------------------------------------------------------
 
+#if __GLASGOW_HASKELL__ < 801
 instance Data TypeRep where
   toConstr _   = error "toConstr"
   gunfold _ _  = error "gunfold"
   dataTypeOf _ = myMkNoRepType "Data.Typeable.TypeRep"
+#endif
 
 
 ------------------------------------------------------------------------------

